### ISSUE TITLE
ProgressBar scrolls instead of redraws when `◎` character is present

### ISSUE THREAD

**Issue by brooksprumo** [console-rs/indicatif#497]
I am experiencing my ProgressBar scrolling instead of truncating/redrawing when I shrink my terminal's window.

I am running on a remote Ubuntu server. Terminal is Mac's iTerm2, with and without tmux. The ProgressBar's template uses `{wide_msg}`, so I expect the progress bar to truncate text as the window sizes decreases.

This only manifests when I shrink my *past* the `◎` character. If this character is not present in the message, the progress bar works as expected regardless of how small I shrink my terminal's window.

Also, version v0.16.2 does *not* exhibit this behavior; it works as expected.

I've tried v0.17.1 and v0.17.2. With v0.17.2 I also tried using the "improved_unicode" feature. None of these resulted in correct behavior.

Here's more information in an issue I have against my repo where I debugged this:
https://github.com/solana-labs/solana/issues/29158

If it's helpful, I can get a screen capture and/or a set of steps to reproduce the issue (may be slightly involved though). Thanks in advance!

**Comment by djc** [comment]
Sorry for regressing this!

Since you already have a working version and a failing version, would you be able to git bisect this? Pinpointing which change broke this would probably make it a bunch easier to diagnose.

**Comment by djc** [comment]
Hmm, are you specifically setting alignment in your templates? It seems like we don't correctly measure the text width in the face of non-ASCII characters when we try to align.

**Comment by djc** [comment]
Right, so the code that trims the output if it's about to overflow the terminal width is wrong. So if `unicode-width` is enabled, the outputs from the `match self.align { .. }` should be a bit more complicated, for example, for `Alignment::Left` we could do something like this:

```
let end = self.str.len() - excess;
if !self.str.is_char_boundary(end) {
    // copy the code for the nightly-only `ceil_char_boundary()`
}
let mut diff = (self.str[..end].width() as i16) - self.width as i16;
while diff != 0 {
    match diff > 0 {
        // trim an additional character off the end of `self.str`
        true => ..
        // increase `end` by one character's worth
        false => ..
    }
}
```

Does that make sense?

Should also add a test to make sure we don't regress this.

**Comment by lgarron** [comment]
I think I might also be running into this with one of my projects. The two messages with an emoji cause the line to overflow by one char.

**Comment by djc** [comment]
I tried to improve the situation with #647 but I can't quite get it to work (and/or my terminal isn't capable enough?). Happy if someone wants to help out (maybe submit a PR against my PR).

**Comment by chris-laplante** [comment]
> I tried to improve the situation with #647 but I can't quite get it to work (and/or my terminal isn't capable enough?). Happy if someone wants to help out (maybe submit a PR against my PR).

I took a peek at this (because I was bored) and I think the issue is that we need to use the grapheme indices, not char indices. I found a crate that already does this which we could use: https://github.com/Aetf/unicode-truncate/tree/master.

FWIW, Windows Terminal 1.22 touts enhanced Unicode support: https://github.com/microsoft/terminal/issues/1472#issuecomment-2242909060. (I noticed this while perusing #611)

I don't really have time to chase this down, but maybe these hints will be helpful to someone else...

**Comment by kojiishi** [comment]
I think there are two independent issues.

The original issue is talking about [U+25CE](https://codepoints.net/U+25CE). It's multi-byte UTF-8, but it's single code unit, it's not as complicated as to need grapheme indices from the Unicode point of view. The issue with this code point is that its "East Asian Width" property value is "ambiguous", meaning that its width is 1 for Latin fonts, but could be 2 for non-Latin (probably East Asian) fonts. The `unicode_width` trait has [`width_cjk()` inaddition to `width()`](https://docs.rs/unicode-width/latest/unicode_width/#cjk-feature-flag) so that callers can determine which to use.

Unfortunately, there's no technically correct way to determine whether these "ambiguous" code point is 1 or 2. What `indicatif` can do is to provide a switch, and callers of `indicatif` should be responsible to determine it (from font etc.).

Another issue is about grapheme cluster consisting of multi-code point characters such as Emoji or combining characters. I found `indicatif` has a feature `improved_unicode` which imports [`unicode-segmentation` trait](https://docs.rs/unicode-segmentation/latest/unicode_segmentation/). This seems promising, but it looks like it's used only for progress chars.

**Comment by djc** [comment]
> Another issue is about grapheme cluster consisting of multi-code point characters such as Emoji or combining characters. I found indicatif has a feature improved_unicode which imports unicode-segmentation trait. This seems promising, but it looks like it's used only for progress chars.

Extending indicatif to use that in more places sounds like a clear win.

**Comment by djc** [comment]
> Unfortunately, there's no technically correct way to determine whether these "ambiguous" code point is 1 or 2. What `indicatif` can do is to provide a switch, and callers of `indicatif` should be responsible to determine it (from font etc.).

And if you have suggestions on how we might implement such a switch, that might also be interesting.

**Comment by kojiishi** [comment]
> > Unfortunately, there's no technically correct way to determine whether these "ambiguous" code point is 1 or 2. What `indicatif` can do is to provide a switch, and callers of `indicatif` should be responsible to determine it (from font etc.).
> 
> And if you have suggestions on how we might implement such a switch, that might also be interesting.

Great, making things more CJK-friendly meets my interests. Created https://github.com/kojiishi/indicatif/pull/1 in my fork as it's on top of PR #799. I can rebase to this repo once #799 is merged, or restructure if #799 is abandoned.

**Comment by kojiishi** [comment]
Correction: I learned PRs can't be moved to a different repo, so I opened #806. Notes on reviewing it is in the PR.
